src/condor_starter.V6.1/starter_v61_main.cpp

```diff
@@ -229,6 +229,81 @@ printClassAd( void )
 	if (!method_list.empty()) {
 		printf("%s = \"%s\"\n", ATTR_HAS_FILE_TRANSFER_PLUGIN_METHODS, method_list.c_str());
 	}
+
+	// even if we have no transfer plugin methods, we may want to
+	// advertise plugin extra attributes
+	if ( ! ft.getPluginQueryAds().empty()) {
+		// publish plugin extra attributes, GetSupportedMethods will collect plugin ads
+		// when it queries the plugins for their '-classad' 
+		// We will incorporate into the starter classad any attributes from the plugins
+		// that are marked as SlotAttrs or StartdAttrs
+		ClassAd extraAd;
+		classad::References failed_methods;
+		classad::References startd_attrs;
+		// when SlotAttrs are enabled, a plugin choose whether to put an attribute
+		// into the daemon ad, or into the slot ads. otherwise all attributes go into all ads
+	   #ifdef ENABLE_PLUGIN_SLOT_ATTRS
+		classad::References slot_attrs;
+	   #endif
+		std::string listval;
+
+		for (auto & ad : ft.getPluginQueryAds()) {
+			classad::References has_attrs;
+			if (ad.LookupString("FailedMethods", listval)) {
+				for (auto & meth : StringTokenIterator(listval)) { failed_methods.insert(meth); }
+			}
+			if (ad.LookupString("StartdAttrs", listval)) {
+				for (auto & attr : StringTokenIterator(listval)) {
+					classad::Value val;
+					if (ad.EvaluateAttr(attr, val)) { // this returns only SAFE values, so no lists or classad
+						extraAd.Insert(attr, classad::Literal::MakeLiteral(val));
+						startd_attrs.emplace(attr);
+						has_attrs.emplace(attr);
+					}
+				}
+			}
+		   #ifdef ENABLE_PLUGIN_SLOT_ATTRS
+			if (ad.LookupString("SlotAttrs", listval)) {
+				for (auto & attr : StringTokenIterator(listval.c_str())) {
+					if ( ! has_attrs.count(attr)) {
+						classad::Value val;
+						if ( ! ad.EvaluateAttr(attr, val)) { // this returns only SAFE values, so no lists or classad
+							continue;
+						} 
+						extraAd.Insert(attr, classad::Literal::MakeLiteral(val));
+					}
+					slot_attrs.emplace(attr);
```

Comment:
Is this is the right spot? To me it seems like all attributes defined in ``SlotAttrs`` will end up in ``slot_attrs`` regardless if ``has_attrs`` contains the attribute or if the attribute is assigned a classad or list

Reply:
Attributes already in has_attrs will end up in SlotAttrs because the earlier loop already decided they were valid. 
Otherwise, they only end up in slot attrs if we never hit the continue because EvaluateAttr did not return false.